clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1163,8 +1251,18 @@ public void handleResponse(AbstractResponse response) {
                 ProducerIdAndEpoch producerIdAndEpoch = new ProducerIdAndEpoch(initProducerIdResponse.data.producerId(),
                         initProducerIdResponse.data.producerEpoch());
                 setProducerIdAndEpoch(producerIdAndEpoch);
-                transitionTo(State.READY);
-                lastError = null;
+                if (!isTransactional()) {
+                    // For idempotent producers, we need to rewrite in-flight batches with the new ID and epoch
+                    topicPartitionBookkeeper.startAllSequencesFromBeginning(producerIdAndEpoch);
```

Comment:
Ok, perhaps this answers one of my questions above. Do we need to rewrite inflight batches necessarily? We may still be able to continue with the initial sequence numbers, which is preferred if possible to avoid introducing duplicates.

Reply:
We are no longer re-writing the sequence numbers of in-flight batches for partitions that do not have an epoch-bump-triggering error